### ISSUE TITLE
DiagnosticListener performance optimization

### ISSUE THREAD

**Issue by lmolkova** [dotnet/corefx#16128]
### Motivation
It is motivated as a performance optimization. The IsEnabled() with no arguments allow a inlinable quick check for performance reasons.

### Details
Add DiagnoticListener.IsEnabled() without arguments as an initial loose check that returns true if there is a subscriber to DiagnosticListener (suggested in #15971).

### API proposal
```C#
public bool IsEnabled() {return _subscriptions != null;};
```

/cc @vancem @stephentoub

**Comment by karelz** [comment]
@vancem if it is ready for API review, please mark it as such before 10am tomorrow (the API review meeting).

**Comment by vancem** [comment]
Note that this is motivated as a performance optimization.    It allows DiagnosticSource code to avoid a virtual calls.   The IsEnabled() with no arguments allow a inlinable quick check, again for performance reasons.

**Comment by karelz** [comment]
@vancem only one api-* label please ;-)

**Comment by karelz** [comment]
@vancem I updated the top post with your 'motivation' - please feel free to make such edits of top post proposal directly in future. Thanks!

**Comment by terrajobst** [comment]
The (1) looks fine.

We're not sure about the argument for sealing. Why do you think the compiler is able to emit a non-virtual call? @jaredpar, is this actually true?

**Comment by vancem** [comment]
When you have a class C and it has a definition, and that method is sealed, then the target MUST be the definition in C (because it must be the most derived version, and sealing it insures that the definition in C is the most derived version).   This means that the method does NOT need to be called virtually (which means it can be inlined)   I have not checked recently but the JIT used to have this exact optimization.

**Comment by benaadams** [comment]
> We're not sure about the argument for sealing. Why do you think the compiler is able to emit a non-virtual call?

Hasn't been true for the 15 year lifetime of .NET; though is often cited. However... its on its way! 🎉 
https://github.com/dotnet/coreclr/pull/9230

**Comment by jaredpar** [comment]
There are two aspects:

1. Can sealing cause the managed compiler to emit a `.call` instruction vs. a `.callvirt`?  The former in some cases is more efficient because it avoids a null check.  In some cases the JIT can avoid the null check even with the `.callvirt`.  But the managed compiler can be more flexible because it only has to obey language semantics while the JIT has to obey IL semantics (stricter).  
1. Can the JIT itself emit more efficient code for a `.callvirt` instruction if the type is sealed?  

As @benaadams points out 2 is not true yet but is on the way in CoreClr.

As for 1 we've been having a lot of discussions about this lately on the managed compiler team.  Met with the JIT team last week to discuss what this would / wouldn't buy us.  The conclusion is that it would likely bring some advantages but we'd need to prototype to see if it was worth the complexity and compat trade offs.

**Comment by benaadams** [comment]
What would be the pattern? If you were holding a `DiagnoticSource` you would need to drop to listener to get the advantages so something like?
```csharp
if((source as DiagnoticListener)?.IsEnabled() ?? true)
{
    // ...
}
```
Or would it be recommended to always hold `DiagnoticListener` directly?

**Comment by lmolkova** [comment]
@benaadams it's recommended to use DiagnosticListener directly.

**Comment by vancem** [comment]
Indeed. the typical usage pattern for DiagnosicSource/DiagnosticListener is for a component (say HTTP) to itself create a static DiagnosticListener (HttpDiagnosticListener), and write to it.   Thus it is very easy for that static variable to be declared as a DiagnosticListener (rather than a DiagnosticSource) and thus the new methods are available.

**Comment by terrajobst** [comment]
```diff
 // Diff between .NET Core 1 and proposed
 namespace System.Diagnostics
 {
     public class DiagnosticListener
     {
+         public bool IsEnabled() {return _subscriptions != null;};
-         public override bool IsEnabled(string name);
+         public sealed override bool IsEnabled(string name); //marked as sealed
+         public sealed override bool IsEnabled(string name, object arg1, object arg2 = null); //marked as sealed
-         public override void Write(string name, object parameters);
+         public sealed override void Write(string name, object parameters); //marked as sealed
     }
 }
 ```

Adding sealed to `IsEnabed(string)` and `Write(string, object)` would be a breaking change to .NET Core 1.0 (source and binary). It's unlikely that we would approve that (@ericstj).

Would you still do the change if you can only seal the newly added methods? We think this would make the API inconsistent and thus even worse. Thoughts?

**Comment by benaadams** [comment]
`public bool IsEnabled()` being non-virtual would get most of the fast "everything-disabled" advantages?

**Comment by lmolkova** [comment]
@benaadams 
Correct, producer should first call IsEnabled without arguments and only then check for a particular event name to be enabled:
```C#
public void OnIncomingRequest(DiagnosticListener listener, HttpContext context)
{
   if (listener.IsEnabled() && listener.IsEnabled("Microsoft.AspNetCore.Hosting.BeginRequest"))
   {
        listener.Write("Microsoft.AspNetCore.Hosting.BeginRequest", new {Context = context})
   }
  //continue processing
}

```

**Comment by lmolkova** [comment]
We can keep all `IsEnabled(string)`, `IsEnabled(string, object, object)` and `Write(string, object)` virtual.

Non-virual IsEnabled without arguments  should be used to keep diagnostics as cheap as possible when off.

@vancem are you ok with it?

**Comment by vancem** [comment]
Go ahead and just add just IsEnabled() as a non-virtual and do nothing else.    

@terrajobst, we knew about the breaking aspect, but it is really quite unlikely (this type has only been public for months, and the use case for extending this clase, let alone overriding this method is hard to come up with), and being CoreCLR (and thus side-by-side) the tolerance for incompatibility is arguably higher.

However given that the value of sealing is lower than the value of having the IsEnabled() call, let's just do the uncontroversial part for now.

**Comment by lmolkova** [comment]
API proposal is changed to IsEnabled() only and ready for review

**Comment by terrajobst** [comment]
Makes sense. Looks good then.

**Comment by stephentoub** [comment]
Closed via #16131
